OpenSim/Common/Path.h

```diff
@@ -0,0 +1,162 @@
+        appendPathElement(pathElement);
+    }
+
+    /// Pure virtual function that returns a char for the designated separator.
+    virtual const char getSeparator() const = 0;
+
+    /// Pure virtual function that returns a string of invalid characters.
+    virtual const std::string getInvalidChars() const = 0;
+
+protected:
+    /// Get an absolute path by resolving it relative to a given otherPath.
+    /// If the current Path is already absolute, return the same Path.
+    std::vector<std::string> formAbsolutePathVec(Path* otherPath);
+
+    /// Find the relative Path between this Path and another Path (otherPath)
+    /// (i.e. the Path to go FROM otherPath TO this Path). Both Paths must be 
+    /// absolute.
+    std::vector<std::string> formRelativePathVec(Path* otherPath);
+
+    /// Return the the sub-path that contains all pathElements except for
+    /// the last one.
+    std::vector<std::string> getParentPathVec()
+    {
+        return getSubPathVec(0, getPathLength() - 1);
+    }
+
+    /// Return the pathElement from the specified position as a string. This
+    /// does not do any checks on the bounds.
+    std::string getPathElement(size_t pos) { return _path[pos]; };
+
+    /// Return the last pathElement as a string.
+    std::string getPathName() { return _path[getPathLength() - 1]; };
+
+private:
+    /// Insert a pathElement at the specified position. Note that this could
+    /// cause a path to become illegal (e.g., adding ".." to the front of 
+    /// an absolute path). The pathElement is checked to ensure no illegal
+    /// characters are used.
+    void insertPathElement(size_t pos, const std::string& pathElement);
+
+    /// Erase a pathElement at the specified position. Note that this could
+    /// cause a path to become illegal (e.g., this may leave a ".." at the
+    /// front of an absolute path).
+    void erasePathElement(size_t pos);
+
+    /// Append a pathElement to the Path, first checking if the pathElement
+    /// is legal.
+    void appendPathElement(const std::string& pathElement);
+
+    /// Cleans up a path. This includes removing "." and resolving ".." if
+    /// possible (i.e. it will not remove leading ".." but otherwise will
+    /// remove the previous pathElement from _path. This method also checks
+    /// if a path is invalid due to an absolute path starting with "..".
+    void cleanPath();
+
+    /// Return the sub-path, on the range [first, last).
+    std::vector<std::string> getSubPathVec(size_t first, size_t last)
+    {
+        std::vector<std::string> subPath(_path.begin() + first,
+            _path.begin() + last);
+        return subPath;
+    }
+
+    /// Return true if pathElement does not contain any chars from the list
+    /// of _invalidChars
+    bool isLegalPathElement(const std::string& pathElement);
+
+    // Path variables
+    std::vector<std::string> _path;
+    char _separator;
+    std::string _invalidChars;
```

Comment:
This could be `const std::string` I think. Not sure if the language will allow it though...